Add a scan for intervals of high depth and excludes reads from those regions from evidence

cwhelan opened the pull request:
This PR attempts to eliminate long-running, useless assemblies that significantly extend runtime on some samples:

- Conducts a scan over the genome to find intervals of excessive depth, defined as an interval where coverage is greater than a lower factor times the average coverage of the sample and containing a coverage peak greater than an upper factor times the average coverage.
- Nearby high-coverage regions within one read-length of each other are merged together.
- Excludes reads that map exclusively inside high coverage regions from evidence gathering.
- Excludes reads that map exclusively inside high coverage regions from QName finding for seeding assemblies.

In addition, after observing that many long-running assemblies occur on non-primary reference contigs, we also exclude reads that map to non-primary contigs (as defined by the "cross-contig to ignore set") from evidence gathering.

Runtime on the CHM mix sample with this change is approximately 38 minutes, and our NA19238 snapshot now takes only 22 minutes, a significant drop in runtime. There are a few changes in the resulting call set but they appear to be minimal.

tedsharpe commented:
It bothers me a bit that we're doing a shuffle (reduceByKey operation in FBES line 880) on the big int arrays of coverage counts.  Would've been so much nicer to process each partition all the way to high-coverage intervals independently, but I understand why it's done this way:  to handle counts that cross partition boundaries.  Since it's a pretty quick step, and since I can't think of a straightforward way to handle partition boundary crossing any better than this, I'm giving it the thumbs up.  I add a few niggles to particular lines and then add another general comment with the approval indication.

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/QNameFinder.java:
```diff
@@ -42,8 +45,11 @@ public QNameFinder( final ReadMetadata metadata,
-        final SVInterval readInterval = new SVInterval(readContigId, readStart, read.getUnclippedEnd()+1);
-        if ( indexedInterval.isDisjointFrom(readInterval) ) return noName;
+        final SVInterval unclippedReadInterval = new SVInterval(readContigId, read.getUnclippedStart(), read.getUnclippedEnd());
+        final SVInterval clippedReadInterval = new SVInterval(readContigId, read.getStart(), read.getEnd());
```
Make no interval before its time.  Move this line below the next one.

cwhelan replied:
done

tedsharpe commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBreakpointEvidenceSpark.java:
```diff
@@ -299,6 +327,30 @@ private static EvidenceScanResults getMappedQNamesSet(
+        final List<SVInterval> depthIntervals = new ArrayList<>((int) ((float) nRefBases / DEPTH_WINDOW_SIZE));
+        for (final SAMSequenceRecord sequenceRecord : header.getSequenceDictionary().getSequences()) {
+            for (int i = 1; i < sequenceRecord.getSequenceLength(); i = i + DEPTH_WINDOW_SIZE) {
+                depthIntervals.add(new SVInterval(readMetadata.getContigID(sequenceRecord.getSequenceName()), i, Math.min(sequenceRecord.getSequenceLength(), i + DEPTH_WINDOW_SIZE)));
```
readMetadata.getContigID(sequenceRecord.getSequenceName()) is a loop invariant (and involves a hashed lookup).  Pull it out as a temporary variable.
(So is sequenceRecord.getSequenceLength(), but that's fast to compute, so it doesn't matter.)

cwhelan replied:
done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBreakpointEvidenceSpark.java:
```diff
@@ -756,47 +808,119 @@ public static IntPair reduce( final IntPair pair1, final IntPair pair2 ) {
             final Logger logger) {
+
+        final int minHighCovFactor = params.highDepthCoverageFactor;
+        final int maxHighCovFactor = params.highDepthCoveragePeakFactor;
```
Renamed these guys to `minFlankingHighCoverageValue` and `minPeakHighCoverageValue` in this method and methods that receive these values as args.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FindBreakpointEvidenceSpark.java:
```diff
@@ -299,6 +327,30 @@ private static EvidenceScanResults getMappedQNamesSet(
+                                                                                  final Logger logger,
+                                                                                  final Broadcast<ReadMetadata> broadcastMetadata) {
+        final long nRefBases = broadcastMetadata.getValue().getNRefBases();
+        final List<SVInterval> depthIntervals = new ArrayList<>((int) ((float) nRefBases / DEPTH_WINDOW_SIZE));
```
Done, went with the exact size calculation, thanks.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/SVReadFilter.java:
```diff
@@ -53,6 +66,19 @@ public boolean isTemplateLenTestable( final GATKRead read ) {
     }
 
+    public boolean containedInRegionToIgnore(final SVInterval interval, final SVIntervalTree<SVInterval> regionsToIgnore) {
+        if (regionsToIgnore.hasOverlapper(interval)) {
```
Done

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/IntervalCoverageFinder.java:
Do you mean `isMappedPrimary` or `isMappedToPrimaryContig`?

cwhelan replied:
You're right, this could be `isMappedToPrimaryContig`. Changed to this to save some unnecessary work.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/IntervalCoverageFinder.java:
done

tedsharpe commented:
Sorry. isMappedToPrimaryContig.
Since we’re not looking for evidence on alts, we probably don’t need info on pile-ups in alts, do we?

cwhelan commented:
Thanks for the comments @tedsharpe , some good code cleanup here.

I've addressed your comments and will merge after tests pass.

I agree about the shuffle -- I'd originally implemented this without it but then found some places where high-depth intervals were getting clipped at the boundaries of depth window partitions due to overlapping counts. In practice it doesn't seem to take a discernible amount of runtime, at least on our development clusters.
